File: app/src/main/java/com/ghstudios/android/features/skills/detail/SkillTreeDetailPagerActivity.java
package com.ghstudios.android.features.skills.detail;

import android.arch.lifecycle.ViewModelProviders;

import com.ghstudios.android.data.DataManager;
import com.ghstudios.android.data.classes.Armor;
import com.ghstudios.android.loader.ItemToSkillTreeListCursorLoader;
import com.ghstudios.android.BasePagerActivity;
import com.ghstudios.android.MenuSection;
import com.ghstudios.android.mhgendatabase.R;

public class SkillTreeDetailPagerActivity extends BasePagerActivity {
    /**
     * A key for passing a monster ID as a long
     */
    public static final String EXTRA_SKILLTREE_ID =
            "com.daviancorp.android.android.ui.detail.skill_id";

    @Override
    public void onAddTabs(TabAdder tabs) {
        long skillTreeId = getIntent().getLongExtra(EXTRA_SKILLTREE_ID, -1);

        SkillDetailViewModel viewModel = ViewModelProviders.of(this).get(SkillDetailViewModel.class);
        viewModel.setSkillTreeId(skillTreeId);

        viewModel.getSkillTreeData().observe(this, (data) -> {
            if (data == null) return;
            setTitle(data.getName());
        });

        tabs.addTab(R.string.skill_tab_detail, () ->
                SkillTreeDetailFragment.newInstance(skillTreeId)
        );

        tabs.addTab(R.string.skill_tab_decorations, () ->
                SkillTreeDecorationFragment.newInstance(skillTreeId,
                        ItemToSkillTreeListCursorLoader.TYPE_DECORATION)
        );

        tabs.addTab(R.string.skill_tab_head, () ->
                SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_HEAD)
        );

        tabs.addTab(R.string.skill_tab_body, () ->
                SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_BODY)
        );

        tabs.addTab(R.string.skill_tab_arms, () ->
                SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_ARMS)
        );

        tabs.addTab(R.string.skill_tab_waist, () ->
                SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_WAIST)
        );

        tabs.addTab(R.string.skill_tab_legs, () ->
                SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_LEGS)
        );
    }

    @Override
    protected int getSelectedSection() {
        return MenuSection.SKILL_TREES;
    }
}

File: app/src/main/java/com/ghstudios/android/features/skills/detail/SkillTreeDetailPagerActivity.kt
package com.ghstudios.android.features.skills.detail

import androidx.lifecycle.Observer
import android.view.Menu
import android.view.MenuItem
import androidx.lifecycle.ViewModelProvider
import com.ghstudios.android.AppSettings

import com.ghstudios.android.data.classes.Armor
import com.ghstudios.android.BasePagerActivity
import com.ghstudios.android.MenuSection
import com.ghstudios.android.mhgendatabase.R

class SkillTreeDetailPagerActivity : BasePagerActivity() {

    companion object {
        /**
         * A key for passing a monster ID as a long
         */
        const val EXTRA_SKILLTREE_ID = "com.daviancorp.android.android.ui.detail.skill_id"
    }

    /**
     * Viewmodel for the entirity of this skill detail, including sub fragments
     */
    private val viewModel by lazy {
        ViewModelProvider(this).get(SkillDetailViewModel::class.java)
    }

    override fun onAddTabs(tabs: BasePagerActivity.TabAdder) {
        val skillTreeId = intent.getLongExtra(EXTRA_SKILLTREE_ID, -1)

        viewModel.setSkillTreeId(skillTreeId, showPenalties=AppSettings.showSkillPenalties)

        viewModel.skillTreeData.observe(this, Observer { data ->
            if (data != null) {
                this.title = data.name
            }
        })

        tabs.addTab(R.string.skill_tab_detail) {
            SkillTreeDetailFragment.newInstance(skillTreeId)
        }

        tabs.addTab(R.string.type_decoration) {
            SkillTreeDecorationFragment.newInstance(skillTreeId)
        }

        tabs.addTab(R.string.skill_tab_head) {
            SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_HEAD)
        }

        tabs.addTab(R.string.skill_tab_body) {
            SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_BODY)
        }

        tabs.addTab(R.string.skill_tab_arms) {
            SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_ARMS)
        }

        tabs.addTab(R.string.skill_tab_waist) {
            SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_WAIST)
        }

        tabs.addTab(R.string.skill_tab_legs) {
            SkillTreeArmorFragment.newInstance(skillTreeId, Armor.ARMOR_SLOT_LEGS)
        }
    }

    override fun getSelectedSection(): Int {
        return MenuSection.SKILL_TREES
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu) // will inflate global actions like search

        menuInflater.inflate(R.menu.menu_skill_detail, menu)
        menu.findItem(R.id.show_negative).isChecked = AppSettings.showSkillPenalties
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle superclass cases first. If handled, return immediately
        val handled = super.onOptionsItemSelected(item)
        if (handled) {
            return true
        }

        return when (item.itemId) {
            R.id.show_negative -> {
                val newSetting = !AppSettings.showSkillPenalties
                AppSettings.showSkillPenalties = newSetting
                viewModel.setShowPenalties(newSetting)
                item.isChecked = newSetting
                true
            }
            else -> false
        }
    }
}
